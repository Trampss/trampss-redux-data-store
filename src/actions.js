const scope = action => `@trampss/${action.toUpperCase()}`

export const SET = prefix => scope(`SET_${prefix}`)
export const set = prefix => payload => ({ type: SET(prefix), payload })

export const RESET = prefix => scope(`RESET_${prefix}`)
export const reset = prefix => () => ({ type: RESET(prefix) })

export const ADD = prefix => scope(`ADD_${prefix}`)
export const add = prefix => payload => ({ type: ADD(prefix), payload })

export const UPDATE = prefix => scope(`UPDATE_${prefix}`)
export const update = prefix => payload => ({ type: UPDATE(prefix), payload })

export const REMOVE = prefix => scope(`REMOVE_${prefix}`)
export const remove = prefix => key => ({ type: REMOVE(prefix), payload: key })

export const ADD_OR_UPDATE = prefix => scope(`ADD_OR_UPDATE_${prefix}`)
export const addOrUpdate = prefix => payload => ({ type: ADD_OR_UPDATE(prefix), payload })
