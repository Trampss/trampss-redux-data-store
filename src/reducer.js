import { keyBy, without, uniq, omit } from 'lodash'
import { SET, ADD, DEL, RESET } from './actions'

export const initState = { data: {}, keys: [], array: [], nb: 0, initialized: false }

export default key => prefix =>
  (state = initState, { type = 'UNKONWN', payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: keyBy(payload, key),
          keys: payload.map(element => element[key]),
          array: payload,
          nb: payload.length,
          initialized: true,
        }
      case ADD(prefix):
        return {
          ...state,
          data: { ...state.data, [payload[key]]: payload },
          keys: uniq([...state.keys, payload[key]]),
          array: [...state.array, payload],
          nb: state.keys.length + 1,
          initialized: true,
        }
      case DEL(prefix):
        return {
          ...state,
          data: omit(state.data, [payload]),
          keys: without(state.keys, payload),
          array: state.array ? state.array.filter(o => o[key] !== payload) : [],
          nb: state.keys.length - 1,
        }
      case RESET(prefix):
        return initState
      default:
        return state
    }
  }
