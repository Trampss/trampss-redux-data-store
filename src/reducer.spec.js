/* eslint-env jest */
import { add } from './actions'
import reducer from './reducer'

const initState = { data: [{ some: 'data' }] }

jest.mock('./middlewares', () => ({
  core: (/* key */) => (/* path */) => ctx => ({
    ...ctx,
    state: { ...ctx.state, core: true, prev: ctx },
    action: { ...ctx.action, core: true },
  }),
}))

const prefix = 'testPrefix'

const middleware = name => key => path => ctx => ({
  state: { ...ctx.state, key, path, [name]: true, prev: ctx },
  action: { ...ctx.action, [name]: true },
})

describe('reducer', () => {
  it('should initialize an action if undefined', () => {
    const testPrefix = reducer({
      engine: [
        middleware('engine'),
      ],
    })('code')(prefix)
    expect(
      testPrefix(initState, undefined),
    ).toMatchSnapshot()
  })

  it('should call -engine- middleware', () => {
    const testPrefix = reducer({
      engine: [
        middleware('engine'),
      ],
    })('code')(prefix)
    expect(
      testPrefix(initState, add(prefix)({ code: '1', some: 'info' })),
    ).toMatchSnapshot()
  })

  it('should call -pre- middlewares', () => {
    const testPrefix = reducer({
      pre: [
        middleware('pre1'),
        middleware('pre2'),
      ],
    })('code')(prefix)

    expect(
      testPrefix(initState, add(prefix)({ code: '1', some: 'info' })),
    ).toMatchSnapshot()
  })

  it('should call -post- middlewares', () => {
    const testPrefix = reducer({
      post: [
        middleware('post1'),
        middleware('post2'),
      ],
    })('code')(prefix)

    expect(
      testPrefix(initState, add(prefix)({ code: '1', some: 'info' })),
    ).toMatchSnapshot()
  })
})
