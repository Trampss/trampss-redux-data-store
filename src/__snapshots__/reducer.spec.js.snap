// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should call -engine- middleware 1`] = `
Object {
  "data": Array [
    Object {
      "some": "data",
    },
  ],
  "defaultData": undefined,
  "engine": true,
  "key": "code",
  "path": "testPrefix",
  "prev": Object {
    "action": Object {
      "payload": Object {
        "code": "1",
        "some": "info",
      },
      "type": "@@krf/ADD_TESTPREFIX",
    },
    "state": Object {
      "data": Array [
        Object {
          "some": "data",
        },
      ],
    },
  },
}
`;

exports[`reducer should call -post- middlewares 1`] = `
Object {
  "data": Array [
    Object {
      "some": "data",
    },
  ],
  "defaultData": undefined,
  "key": "code",
  "path": "testPrefix",
  "post1": true,
  "post2": true,
  "prev": Object {
    "action": Object {
      "payload": Object {
        "code": "1",
        "some": "info",
      },
      "post1": true,
      "type": "@@krf/ADD_TESTPREFIX",
    },
    "state": Object {
      "data": Array [
        Object {
          "some": "data",
        },
      ],
      "defaultData": undefined,
      "key": "code",
      "path": "testPrefix",
      "post1": true,
      "prev": Object {
        "action": Object {
          "payload": Object {
            "code": "1",
            "some": "info",
          },
          "type": "@@krf/ADD_TESTPREFIX",
        },
        "state": Object {
          "data": Array [
            Object {
              "some": "data",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`reducer should call -pre- middlewares 1`] = `
Object {
  "data": Array [
    Object {
      "some": "data",
    },
  ],
  "defaultData": undefined,
  "key": "code",
  "path": "testPrefix",
  "pre1": true,
  "pre2": true,
  "prev": Object {
    "action": Object {
      "payload": Object {
        "code": "1",
        "some": "info",
      },
      "pre1": true,
      "type": "@@krf/ADD_TESTPREFIX",
    },
    "state": Object {
      "data": Array [
        Object {
          "some": "data",
        },
      ],
      "defaultData": undefined,
      "key": "code",
      "path": "testPrefix",
      "pre1": true,
      "prev": Object {
        "action": Object {
          "payload": Object {
            "code": "1",
            "some": "info",
          },
          "type": "@@krf/ADD_TESTPREFIX",
        },
        "state": Object {
          "data": Array [
            Object {
              "some": "data",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`reducer should initialize an action if undefined 1`] = `
Object {
  "data": Array [
    Object {
      "some": "data",
    },
  ],
  "defaultData": undefined,
  "engine": true,
  "key": "code",
  "path": "testPrefix",
  "prev": Object {
    "action": Object {
      "payload": undefined,
      "type": "UNKNOWN",
    },
    "state": Object {
      "data": Array [
        Object {
          "some": "data",
        },
      ],
    },
  },
}
`;
